### ISSUE TITLE
Phase A — Validate feat/lstm-kl-imitation PR (LSTM + KL + imitation-init)

### ISSUE THREAD

**Issue by aberson** [aberson/Alpha4Gate#100]
## Goal

Validate the `feat/lstm-kl-imitation` PR (commit `498f405`) does not regress baseline performance before any subsequent AlphaStar-upgrade phase is layered on. This is the blocking first phase of `documentation/plans/alphastar-upgrade-plan.md`.

## Branch

`feat/lstm-kl-imitation` — adds `MlpLstmPolicy` (RecurrentPPO), KL-to-rules auxiliary loss, and imitation-init. All flags default to off in `data/hyperparams.json` so master behavior is unchanged until flipped.

## Prerequisites

None. First phase of the AlphaStar-upgrade work.

## Scope

Six sequential validation steps. Do not batch — the purpose is to isolate regressions per feature.

### A.0 — Branch checkout + unit tests

```powershell
git checkout feat/lstm-kl-imitation
uv run pytest -q
```

Pass: 834 tests green (7 pre-existing failures in test_api replay endpoints and test_rewards unrelated to this PR).

### A.1 — No-op regression (all new flags at safe defaults)

```powershell
uv run python -m alpha4gate.runner --train rl --cycles 1 --games-per-cycle 3 --difficulty 3
```

Pass: cycle completes, checkpoint saves, win rate logged. Fail: any import / shape / class-mismatch error from the dispatch logic in `_init_or_resume_model`.

### A.2 — Imitation init alone

```powershell
(Get-Content data/hyperparams.json) -replace '"use_imitation_init": false', '"use_imitation_init": true' | Set-Content data/hyperparams.json
uv run python -m alpha4gate.runner --train rl --cycles 1 --games-per-cycle 3 --difficulty 3 --ensure-pretrain
```

Pass: log shows `--ensure-pretrain: running imitation training` → `agreement=X.XXX` → `Loading imitation-pretrained checkpoint v0_pretrain` before cycle 1. `Test-Path data/checkpoints/v0_pretrain.zip` returns True.

Re-run without `--ensure-pretrain` to confirm idempotence:

```powershell
uv run python -m alpha4gate.runner --train rl --cycles 1 --games-per-cycle 3 --difficulty 3
```

Pass: log shows `Loading imitation-pretrained checkpoint v0_pretrain` without re-running imitation.

### A.3 — KL-to-rules alone

```powershell
(Get-Content data/hyperparams.json) `
  -replace '"use_imitation_init": true', '"use_imitation_init": false' `
  -replace '"kl_rules_coef": 0.0', '"kl_rules_coef": 0.1' `
  | Set-Content data/hyperparams.json

uv run python -m alpha4gate.runner --train rl --cycles 2 --games-per-cycle 3 --difficulty 3
```

Pass: no crash, cycle wall-clock ≤ 1.5× A.1 wall-clock (extra pass overhead bounded). Optional: `data/training_diagnostics.json` probabilities on diagnostic states drift toward rule-engine choice across cycles.

### A.4 — LSTM alone

```powershell
(Get-Content data/hyperparams.json) `
  -replace '"kl_rules_coef": 0.1', '"kl_rules_coef": 0.0' `
  -replace '"policy_type": "MlpPolicy"', '"policy_type": "MlpLstmPolicy"' `
  | Set-Content data/hyperparams.json

# LSTM checkpoint is incompatible with prior MlpPolicy checkpoints
Move-Item data/checkpoints data/checkpoints.bak-pre-lstm

uv run python -m alpha4gate.runner --train rl --cycles 2 --games-per-cycle 3 --difficulty 3
```

Pass: env loop runs, hidden state threads through, cycles complete.

Known failure: if `net_arch: [128, 128]` flat-list isn't valid for `MlpLstmPolicy`, model construction crashes. Fix — change `net_arch` to `{"pi": [128], "vf": [128]}` and log it in the plan history.

### A.5 — All three together

```powershell
(Get-Content data/hyperparams.json) `
  -replace '"use_imitation_init": false', '"use_imitation_init": true' `
  -replace '"kl_rules_coef": 0.0', '"kl_rules_coef": 0.1' `
  | Set-Content data/hyperparams.json

uv run python -m alpha4gate.runner --train rl --cycles 3 --games-per-cycle 3 --difficulty 3 --ensure-pretrain
```

### A.6 — Validation soak

```powershell
uv run python -m alpha4gate.runner --batch 20 --difficulty 3 --decision-mode hybrid `
  --model-path data/checkpoints/best.zip
```

## Known-failure diagnostic table

| Symptom | Likely cause | Fix |
|---------|--------------|-----|
| `lstm_states` / `RNNStates` error at model build | `net_arch: [128,128]` flat, LSTM wants dict | Change to `{"pi":[128],"vf":[128]}` |
| `cross_entropy` NaN during KL pass | coef too high or obs decode broken | Drop to 0.05; probe `rule_actions_for_batch` output |
| Imitation agreement stuck ≈ 0.17 (1/6) | DB transitions empty or all one action | `uv run python -c "from alpha4gate.learning.database import TrainingDB; db=TrainingDB('data/training.db'); print(db.get_transition_count())"` |
| Trainer loads wrong class on resume | Pre-existing checkpoint from different `policy_type` | Clear `data/checkpoints/` or start fresh dir |

## Gate (pass criteria)

At least one of the flag combos in A.2–A.5 produces win rate ≥ baseline (75% at difficulty 3) over 20 validation games (A.6), with no crashes and all unit tests green. On pass: merge branch to master, tag `alphastar/A/final`.

## Kill criterion

All four configurations regress >10% win rate vs baseline. Root-cause candidates:

- Stateless rule-teacher too lossy (see `rules_policy.py` "TRADE-OFF" block). Upgrade to option (b) from the PR conversation — record `rule_action` per step in `SC2Env`.
- Imitation padding introduces distribution shift overriding PPO signal. Disable padding and use `BASE_GAME_FEATURE_DIM`-only obs for RL.

## Rollback

Branch unmerged until gate passes → rollback is `git checkout master`. Post-merge regret: `git revert` the merge commit.

## Reference

Full context, including phases B–F that depend on this one: [`documentation/plans/alphastar-upgrade-plan.md`](../blob/master/documentation/plans/alphastar-upgrade-plan.md).

---

*Phase A of AlphaStar-inspired upgrade plan. Umbrella issue tracking validation of PR `feat/lstm-kl-imitation` (commit 498f405).*

**Comment by aberson** [comment]
## build-phase: Phase A orchestration started

Runbook: `documentation/plans/phase-a-buildphase-runbook.md`
Total steps: 8 (7 operator, 1 wait)

### Step 1 started — Checkout feat/lstm-kl-imitation and confirm baseline tests

**Type:** operator (manual)
**Started:** 2026-04-14

Orchestrator halted, awaiting operator action. Working tree is currently dirty on master (modified `data/reward_rules.json`, modified `.claude/skills/improve-bot-advised/SKILL.md`, and several untracked data/doc dirs). Step 1 will commit/stash those, check out the branch, and run `uv run pytest`.

**Comment by aberson** [comment]
Plan merged 2026-04-15. This phase's scope is unchanged, but the canonical plan file moved:

- **Old:** `documentation/plans/alphastar-upgrade-plan.md` or `documentation/plans/bot-versioning-selfplay-plan.md` (both now archived under `documentation/archived/`)
- **New:** [documentation/plans/alpha4gate-master-plan.md](../blob/master/documentation/plans/alpha4gate-master-plan.md)

Umbrella #105 retitled to the master plan.

**Comment by aberson** [comment]
## build-phase: Phase A resumed

**Plan:** documentation/plans/phase-a-buildphase-runbook.md
**Branch:** feat/lstm-kl-imitation @ 6d741d6 (master merged in as 68a27f7)

### Step 1 complete

- **Verdict:** PASS (operator)
- **Evidence:** 864 unit tests passing on the merged branch. Master had drifted 2 days ahead; merge `68a27f7` resolved with no conflicts.
- **Plan updated:** Status: DONE marked in runbook.

### Step 2 starting — A.1 no-op regression

Operator step. Awaiting user execution of:
``````
uv run python -m alpha4gate.runner --train rl --cycles 1 --games-per-cycle 3 --difficulty 3
``````
With shipped defaults (`use_imitation_init: false`, `kl_rules_coef: 0.0`, `policy_type: MlpPolicy`).

**Pass criteria:** cycle completes, checkpoint saves, win-rate logged, no error from `_init_or_resume_model`. Wall-clock recorded for use as Step 4 baseline.

**Comment by aberson** [comment]
## build-phase: Step 2 complete — A.1 no-op regression PASS

**Verdict:** PASS
**Completion:** 2026-04-15 16:17:45
**Checkpoint saved:** `data/checkpoints/v1.zip` (best=False)
**Win rate:** 0/3 (expected — regression test, not performance)
**Failed games:** 0
**Wall-clock baseline:** ~20 min for 3 games at diff 3 (Step 4 KL-overhead gate → ≤30 min)

No `_init_or_resume_model` error. New Phase A flags at shipped defaults are backward-compatible with the existing MlpPolicy training loop.

### Step 3 starting — A.2 imitation-init alone

Operator step. Flipping `use_imitation_init: true` in `data/hyperparams.json` (kl_rules_coef=0.0, policy_type=MlpPolicy). Running with `--ensure-pretrain`.

**Pass criteria:**
- Log shows `--ensure-pretrain: running imitation training` → `agreement=X.XXX` → `Loading imitation-pretrained checkpoint v0_pretrain` before cycle 1
- `data/checkpoints/v0_pretrain.zip` exists
- Re-run without `--ensure-pretrain` is idempotent (loads without re-running imitation)

Likely suspect if blocked: empty training.db or skewed action distribution (see diagnostic table in `documentation/archived/alphastar-upgrade-plan.md`).

**Comment by aberson** [comment]
## build-phase: Step 3 complete — A.2 imitation-init PASS

**Verdict:** PASS (after one fix)

### Imitation training result
- **Agreement: 0.915** (well above the 0.17 feared in the diagnostic table)
- 100 epochs, final loss 0.2827
- 94,650 transitions consumed
- `data/checkpoints/v0_pretrain.zip` saved (best=True)

### Idempotence confirmed
Second run with `use_imitation_init: true` in hyperparams (no `--ensure-pretrain` flag):
``````
17:26:41,067 alpha4gate.learning.trainer INFO Loading imitation-pretrained checkpoint v0_pretrain (class=PPO)
17:26:42,167 alpha4gate.learning.trainer INFO === Cycle 1/1 (difficulty=3) ===
``````
No re-run of imitation training. Pretrain loads in ~1s.

### Pre-existing DB bug fixed mid-step (commit `d37fdfe`)
`sample_batch` crashed with `invalid literal for int()` on historical training.db rows. Root cause: sqlite3's default adapter stores numpy `int64` scalars as 8-byte BLOBs via buffer protocol. The `action` column had ~40K blob rows interleaved with ~54K integer rows across DB history. Fix:
- **Write path:** `int(action)` + `float(reward)` before INSERT (prevents new blobs)
- **Read path:** `_coerce_action` decodes little-endian int64 from blob rows (preserves historical data)

All 864 unit tests still pass. Bug is pre-existing (not from this branch); the new `sample_batch` consumer is the first to exercise the read path.

### Step 4 starting — A.3 KL-to-rules alone
Flipping hyperparams back:
- `use_imitation_init: false`
- `kl_rules_coef: 0.1`
- `policy_type: MlpPolicy`

Will run 2 cycles × 3 games. Per-cycle wall-clock must be ≤ 442 sec (1.5× of Step 2's 295-sec baseline). Watching for NaN crashes in the KL pass.

**Comment by aberson** [comment]
## build-phase: Step 4 complete — A.3 KL-to-rules PASS

**Verdict:** PASS
**Wall-clock:** 502 sec for 2 cycles → **251 sec/cycle** (Step 2 baseline: 295 sec/cycle)
**Ratio:** 0.85× baseline (gate: ≤1.5× → **442 sec**; cleared with room)
**NaN:** none
**Checkpoints saved:** v1, v2 (both cycles)
**Failed games:** 0/6

`kl_rules_coef=0.1` adds no measurable per-step overhead on this bot's hot path.

### Step 5 starting — A.4 LSTM alone

Operator step. Config:
- `use_imitation_init: false`
- `kl_rules_coef: 0.0`
- `policy_type: MlpLstmPolicy`

First: backup prior checkpoints (`Move-Item data/checkpoints data/checkpoints.bak-pre-lstm`) since MlpLstmPolicy checkpoints are incompatible with MlpPolicy checkpoints.

**Known failure mode:** `net_arch: [128, 128]` flat list is invalid for MlpLstmPolicy. If model construction crashes, the fix is to change `net_arch` in hyperparams to `{"pi": [128], "vf": [128]}`. Any net_arch fix will be logged in the plan history section.

**Comment by aberson** [comment]
## build-phase: Step 5 complete — A.4 LSTM PASS

**Verdict:** PASS
**Wall-clock:** 543 sec for 2 cycles (~271 sec/cycle)
**Checkpoints:** v1, v2 saved (MlpLstmPolicy class)
**NaN:** none
**Failed games:** 0/6

**Unexpected result:** `RecurrentPPO` accepted `net_arch: [128, 128]` flat list without crash. The runbook's predicted failure mode (requiring rewrite to `{"pi": [128], "vf": [128]}` dict form) did not materialize — no hyperparams patch needed.

### Step 6 starting — A.5 all three together

Operator step. Full stack:
- `use_imitation_init: true`
- `kl_rules_coef: 0.1`
- `policy_type: MlpLstmPolicy` (keep)

Will run 3 cycles × 3 games with `--ensure-pretrain`. This exercises whether imitation pretrain can be rebuilt for the LSTM policy class and loaded cleanly into `RecurrentPPOWithKL`. Previous `v0_pretrain.zip` is in the pre-lstm backup dir, so `--ensure-pretrain` will regenerate it for the current policy class.

Pass: 3 cycles complete, imitation agreement logged, pretrain loads cleanly, no NaN in KL pass.

**Comment by aberson** [comment]
## build-phase: Step 6 complete — A.5 full stack PASS

**Verdict:** PASS
**End:** 2026-04-15 18:24:19
**Cycles:** 3/3 completed, checkpoints v1/v2/v3 saved
**Per-cycle WR:** 0.0, 0.167, 0.167 (sliding 6-game window)
**Total:** ~1 win / 9 games (expected — 9 games is not enough for the policy to learn)
**NaN:** none
**Failed games:** 0/9

The critical unknown cleared: `RecurrentPPOWithKL` loaded the `RecurrentPPO`-class `v0_pretrain.zip` without serialization issues. imitation + KL + LSTM all compose cleanly.

### Step 5 on the unit-test front
- 864 tests still green after d37fdfe DB blob fix
- All Phase A config variations completed end-to-end

### Step 7 — HALTING FOR WAIT STEP

Step 7 (A.6) is the 20-game validation soak at difficulty 3 hybrid. That's ~2–4 hours of SC2 wall-clock that shouldn't burn context window — build-phase is handing off.

**User action:** when ready, run the soak manually per the plan. Resume build-phase in a fresh session with:

``````
/build-phase --plan documentation/plans/phase-a-buildphase-runbook.md --resume 8
``````

(Step 7 stays marked pending; mark it DONE manually in the runbook when the soak wraps, or I'll do it on resume given the batch log.)

**Comment by aberson** [comment]
## Phase A COMPLETE — merged to master

**Tag:** `alphastar/A/final` on `cfeeb99`
**Merge:** fast-forward, 17 files changed, +720 / -108 lines

### Gate evaluation
| Criterion | Threshold | Result |
|-----------|-----------|--------|
| combo_passed (20-game soak WR) | ≥ 75% | **95% (19/20)** |
| no_crashes (Steps 2–6) | all pass | **6/6 PASS** |
| tests_green | 864 passing | **864/864** |

### Validation summary
| Step | Test | Verdict | Key metric |
|------|------|---------|------------|
| 1 (A.0) | Baseline tests | PASS | 864 tests on merged branch |
| 2 (A.1) | No-op regression | PASS | 295 sec/cycle, no errors |
| 3 (A.2) | Imitation-init | PASS | agreement=0.915, idempotent load |
| 4 (A.3) | KL-to-rules | PASS | 0.85× baseline overhead |
| 5 (A.4) | LSTM alone | PASS | flat net_arch accepted (unexpected) |
| 6 (A.5) | Full stack | PASS | 3 cycles, no NaN, no class-mismatch |
| 7 (A.6) | 20-game soak | PASS | **19/20 = 95% WR** at diff 3 hybrid |
| 8 | Gate + merge | PASS | merged, tagged, pushed |

### Bugs found and fixed during validation
1. **DB blob fix** (`d37fdfe`): pre-existing bug where numpy int64 was stored as BLOB in training.db action column. Fixed write path (int coercion) and read path (little-endian decoder). Not introduced by this branch.
2. **Advisor CancelledError in batch mode**: `claude_advisor._call_api` CancelledError in `--batch --decision-mode hybrid`. Workaround: `--no-claude`. Filed as separate follow-up.

### New code shipped (17 files)
- `ppo_kl.py` — PPOWithKL / RecurrentPPOWithKL (KL-to-rules regularization)
- `rules_policy.py` — stateless rule-engine teacher for KL target
- `policy_probe.py` — diagnostic probing for imitation agreement
- Enhanced `imitation.py` — policy-type-aware BC with recurrent support
- Enhanced `trainer.py` — imitation-init + KL + LSTM dispatch in `_init_or_resume_model`
- Enhanced `runner.py` — `--ensure-pretrain` flag
- 4 new test files (221 new test lines)

### What's next
Per master plan: **Phase 1** — move code into `bots/v0/` isolation. Phase A proved the features work; Phase 1 packages them for the versioned bot framework.

Full-stack hyperparams stashed (`git stash list` → "phase-A full-stack hyperparams") for future use.
